store/store.go

```diff
@@ -233,6 +240,125 @@ func (s *SnapUbuntuStoreRepository) checkStoreResponse(resp *http.Response) {
 	}
 }
 
+/*
+purchase encapsulates the purchase data sent to us from the software center agent.
+
+This object type can be received in response to requests about the user's current
+purchases, and also when making purchase requests.
+
+When making a purchase request, the State "InProgress", together with a RedirectTo
+URL may be received. In-this case, the user must be directed to that webpage.
+Additionally, Partner ID may be recieved as an extended header "X-Partner-Id",
+this should be included in the follow-on requests to the redirect URL.
+
+HTTP/1.1 200 OK
+Content-Type: application/json; charset=utf-8
+
+[
+  {
+    "open_id": "https://login.staging.ubuntu.com/+id/open_id",
+    "package_name": "com.ubuntu.developer.dev.appname",
+    "refundable_until": "2015-07-15 18:46:21",
+    "state": "Complete"
+  },
+  {
+    "open_id": "https://login.staging.ubuntu.com/+id/open_id",
+    "package_name": "com.ubuntu.developer.dev.appname",
+    "item_sku": "item-1-sku",
+    "purchase_id": "1",
+    "refundable_until": null,
+    "state": "Complete"
+  },
+  {
+    "open_id": "https://login.staging.ubuntu.com/+id/open_id",
+    "package_name": "com.ubuntu.developer.dev.otherapp",
+    "refundable_until": "2015-07-17 11:33:29",
+    "state": "Complete"
+  }
+]
+*/
+type purchase struct {
+	OpenID          string `json:"open_id"`
+	SnapID          string `json:"snap_id"`
+	RefundableUntil string `json:"refundable_until"`
+	State           string `json:"state"`
+	ItemSKU         string `json:"item_sku,omitempty"`
+	PurchaseID      string `json:"purchase_id,omitempty"`
+	RedirectTo      string `json:"redirect_to,omitempty"`
+}
+
+type purchaseList []*purchase
+
+func (s *SnapUbuntuStoreRepository) getPurchasesFromURL(url *url.URL, channel string, auther Authenticator) (purchaseList, error) {
+	if auther == nil {
+		return nil, fmt.Errorf("cannot obtain known purchases from store: no authentication credentials provided")
+	}
+
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	s.applyUbuntuStoreHeaders(req, "", auther)
+	req.Header.Set("X-Ubuntu-Device-Channel", channel)
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var purchases purchaseList
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		dec := json.NewDecoder(resp.Body)
+		if err := dec.Decode(&purchases); err != nil {
+			return nil, fmt.Errorf("cannot decode known purchases from store: %v", err)
+		}
+	case http.StatusNotFound:
+		// the store returns 404 for snaps that have no purchases, so just return an empty list
```

Comment:
@niemeyer have sent a request to the server team to see if we can improve this situation